File: app/src/main/java/ru/atomofiron/regextool/screens/preferences/PrefsFragment.java
package ru.atomofiron.regextool.screens.preferences;

import android.content.SharedPreferences;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.preference.Preference;
import androidx.preference.PreferenceCategory;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceScreen;
import androidx.preference.SwitchPreferenceCompat;

import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;

import ru.atomofiron.regextool.Util;
import ru.atomofiron.regextool.R;
import ru.atomofiron.regextool.utils.Cmd;

public class PrefsFragment extends PreferenceFragmentCompat implements Preference.OnPreferenceChangeListener {
	public static final ArrayList<String> changedPrefs = new ArrayList<>();

	private SharedPreferences sp;

	public PrefsFragment() {}

	@Override
	public void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		addPreferencesFromResource(R.xml.preferences);
		setHasOptionsMenu(true);

		sp = Util.sp(getActivity());
	}

	@Override
	public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
		super.onCreateOptionsMenu(menu, inflater);
		inflater.inflate(R.menu.menu_preferences, menu);
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		if (item.getItemId() == R.id.tips)
			Util.showHelp(getContext());

		return super.onOptionsItemSelected(item);
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		PreferenceScreen prefScreen = getPreferenceScreen();
		for (int i = 0; i < prefScreen.getPreferenceCount(); i++) {
			Preference pref = prefScreen.getPreference(i);
			if (pref.getClass().equals(PreferenceCategory.class)) {
				PreferenceCategory prefCategory = (PreferenceCategory) pref;
				for (int j = 0; j < prefCategory.getPreferenceCount(); j++) {
					pref = prefCategory.getPreference(j);
					pref.setOnPreferenceChangeListener(this);
					update(pref, null);
				}
			}
		}
		return super.onCreateView(inflater, container, savedInstanceState);
	}

	@Override
	public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
	}

	@Override
	public boolean onPreferenceChange(Preference preference, Object newValue) {
		boolean result = update(preference, newValue);

		if (preference.getKey().equals(Util.PREF_THEME) || preference.getKey().equals(Util.PREF_ORIENTATION))
			getActivity().recreate();
		else if (preference.getKey().equals(Util.PREF_MAX_SIZE))
			update(preference, newValue);

		// чтобы можно было проверить факт изменения определённой конфигурации
		// и совершить необходимые действия
		String key = preference.getKey();
		switch (key) {
			case Util.PREF_SPECIAL_CHARACTERS:
				changedPrefs.add(key);
		}

		return result;
	}

	private boolean update(Preference pref, Object newValue) {
		String value = newValue == null ? null : newValue.toString();

		String key = pref.getKey();
		switch (key) {
			case Util.PREF_STORAGE_PATH:
			case Util.PREF_EXTRA_FORMATS:
				pref.setSummary(value == null ? sp.getString(key, "") : value);
				break;
			case Util.PREF_SPECIAL_CHARACTERS:
				pref.setSummary(value == null ?
						sp.getString(Util.PREF_SPECIAL_CHARACTERS, Util.DEFAULT_SPECIAL_CHARACTERS) : value);
				break;
			case Util.PREF_THEME:
				pref.setSummary(getResources().getStringArray(R.array.theme_var)
						[Integer.parseInt(value == null ? sp.getString(key, "0") : value)]);
				break;
			case Util.PREF_ORIENTATION:
				int i = Integer.parseInt(value == null ? sp.getString(key, "2") : value);
				pref.setSummary(getResources().getStringArray(R.array.orientation_var)[i]);
				break;
			case Util.PREF_USE_SU:
				if (value == null) {
					if (sp.getBoolean(key, false) && !Cmd.checkSu())
						((SwitchPreferenceCompat)pref).setChecked(false);
				} else if (value.equals("true"))
					return Cmd.checkSu();

				break;
			case Util.PREF_MAX_SIZE:
				if (newValue == null)
					newValue = sp.getInt(key, 0);

				pref.setSummary(Util.intToHumanReadable((int) newValue, getResources().getStringArray(R.array.size_suffix_arr)));
				break;
		}
		return true;
	}

}

File: app/src/main/java/ru/atomofiron/regextool/screens/preferences/PrefsFragment.kt
package ru.atomofiron.regextool.screens.preferences

import android.os.Bundle
import android.view.*
import androidx.preference.Preference
import androidx.preference.PreferenceCategory
import androidx.preference.PreferenceFragmentCompat
import ru.atomofiron.regextool.R
import ru.atomofiron.regextool.iss.store.SettingsStore
import ru.atomofiron.regextool.utils.Const
import ru.atomofiron.regextool.utils.Shell.checkSu
import ru.atomofiron.regextool.utils.Util
import java.util.*

class PrefsFragment : PreferenceFragmentCompat(), Preference.OnPreferenceChangeListener {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        addPreferencesFromResource(R.xml.preferences)
        setHasOptionsMenu(true)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_preferences, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.tips) {
            Util.showHelp(context)
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val prefScreen = preferenceScreen
        for (i in 0 until prefScreen.preferenceCount) {
            var pref = prefScreen.getPreference(i)
            if (pref.javaClass == PreferenceCategory::class.java) {
                val prefCategory = pref as PreferenceCategory
                for (j in 0 until prefCategory.preferenceCount) {
                    pref = prefCategory.getPreference(j)
                    pref.onPreferenceChangeListener = this
                    updatePreference(pref, null)
                }
            }
        }
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) = Unit

    override fun onPreferenceChange(preference: Preference, newValue: Any): Boolean {
        return updatePreference(preference, newValue)
    }

    private fun updatePreference(pref: Preference, newValue: Any?): Boolean {
        when (pref.key) {
            Const.PREF_STORAGE_PATH -> {
                pref.summary = newValue as? String
                if (newValue is String) {
                    SettingsStore.storagePath.notify(newValue)
                }
            }
            Const.PREF_EXTRA_FORMATS -> {
                pref.summary = newValue as? String
                if (newValue is String) {
                    SettingsStore.extraFormats.notify(newValue)
                }
            }
            Const.PREF_SPECIAL_CHARACTERS -> {
                pref.summary = newValue as? String
                if (newValue is String) {
                    SettingsStore.specialCharacters.notify(newValue)
                }
            }
            Const.PREF_APP_THEME -> {
                val i = newValue as? Int ?: SettingsStore.appTheme.value.ordinal
                pref.summary = resources.getStringArray(R.array.theme_var)[i]
                if (newValue is Int) {
                    SettingsStore.appTheme.notifyByOriginal(i)
                }
            }
            Const.PREF_APP_ORIENTATION -> {
                val i = newValue as? Int ?: SettingsStore.appOrientation.value.ordinal
                pref.summary = resources.getStringArray(R.array.orientation_var)[i]
                if (newValue is Int) {
                    SettingsStore.appOrientation.notifyByOriginal(i)
                }
            }
            Const.PREF_USE_SU -> {
                val use = newValue as? Boolean ?: SettingsStore.useSu.value
                val allow = !use || checkSu()
                if (allow && newValue is Boolean) {
                    SettingsStore.useSu.notify(newValue)
                }
                return allow
            }
            Const.PREF_MAX_SIZE -> {
                val maxSize = newValue ?: SettingsStore.maxFileSizeForSearch.value
                val view = view
                if (view != null) {
                    val intValue = maxSize as Int
                    view.post(Runnable {
                        // иначе вьюха меняется в процессе рассчётов списка
                        val suffixes = resources.getStringArray(R.array.size_suffix_arr)
                        pref.summary = Util.intToHumanReadable(intValue, suffixes)
                    })
                }
                if (newValue is Int) {
                    SettingsStore.maxFileSizeForSearch.notify(newValue)
                }
            }
        }
        return true
    }
}